### PR TITLE
Use torchnet

### PR DESCRIPTION
**Pull request by szagoruyko**
- get rid of augmentation.lua, preprocess data in several threads (and everything in one file which is nice)
- CPU support

Have tested on a couple of small networks, will merge later after more testing.

### CODE REVIEW

**Comment by dlmacedo** [comment]
Dear szagoruyko,

Which piece of code the fastest option is redundant with?

Will we be able to disable it? How?

Thanks,

David

**Comment by dlmacedo** [comment]
Dear Sergey Zagoruyko,

I guess if we set "cudnn_determinist = true" we get something similar to old "cudnn_fastest = false".

If we set  "cudnn_determinist = false" we get something similar to old "cudnn_fastest = true".

Am I right?

David

**Comment by szagoruyko** [comment]
@dlmacedo it cudnn.fastest is redundant when cudnn.benchmark = true

File: train.lua [hunk 1]
```diff
@@ -87,98 +79,144 @@ do
...
+                  input = dataset.data[idx]:float(),
+                  target = torch.LongTensor{dataset.labels[idx]},
+               }
+            end,
+         }
+
+         return tnt.BatchDataset{
+            batchsize = opt.batchSize,
+            policy = 'skip-last',
+            dataset = list_dataset,
```
Comment (fmassa) [inline train.lua]: redundant line?

File: train.lua [hunk 2]
```diff
@@ -87,98 +79,144 @@ do
...
-function train()
-  model:training()
-
-  local targets = torch.CudaTensor(opt.batchSize)
-  local indices = torch.randperm(provider.trainData.data:size(1)):long():split(opt.batchSize)
-  -- remove last element so that all minibatches have equal size
-  indices[#indices] = nil
-
-  local loss = 0
+local function randomcrop(x)
```
Comment (fmassa) [inline train.lua:90]: you could maybe make this function return a function that performs the crop, but using as upvalues the module? This way you avoid some branching due to the `if .. else`, and would also avoid generating a new module at each data access

Comment (szagoruyko) [inline train.lua:90]: I think the overhead is small for now, will wait for torchnet-vision to fix this

**Pull request merged**